### PR TITLE
Feature/build compatibles

### LINKED ISSUE #14291: [feature] conan install --build should support building packages that are only compatible and not an exact match
### What is your suggestion?

The new compatibility function allows `conan install` to use compatible packages instead of an exact match. However, I think it should work together with the `--build` parameter to force building packages that are only compatible (but use e.g. an older C++ standard or an older libc version). A more exact matching package might have performance or other advantages over using only a compatible package.

It could be called `--build compatible` or something similar.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

### LINKED ISSUE #16148: [feature] easy fallback at recipe create/build when cppstd is not invalid for building, but not for consuming
### What is your suggestion?

For some libraries, we may have the following situation:

A) The library uses C++ features that have been removed from newer standards, i.e. it can be built with c++14 but *not* with c++17)
B) This uses of removed functionality are used _only_ in implementation files - and not exposed in the interface, i.e., a consumer can be in C++20, depend on said library that was built for C++14, and this is fine (compiler vendors make a lot of efforts to retain binary compatibility across C++ standards)

For (A), the recommended way in Conan 2.0 to reflect this is the following:
```
def validate_build(self):
     check_max_cppstd(self, "14")
```

This will prevent Conan from _creating_ the binary with a cppstd higher than 17.

For (B), the fallback is in the `compatibility.py` plugin - Conan will try to find "compatible" binaries, that is, if we request `compiler.cppstd=17`, it will _not_ find a binary for the library matching the configuration, but it will happily use the binary with `compiler.cppstd=14` if it exists. 

But the fallback in `compatibility.py` only works if the compatible binary already exists. If it doesn't, and one has `--build=missing`, it will fail during `validate_build()`, and the user has to explicitly express that they want `cppstd=14`. But this is not always intuitive: it would have to be `-s "library/*:compiler.cppstd=14"`, otherwise it would affect all binaries, e.g. if there are dependencies. 

It's a similar case during Conan create. Can we consider a fallback, similar to `compatibility.py` at consumption time, but at "creation" time?  

### Have you read the CONTRIBUTING guide?

- [ ] I've read the CONTRIBUTING guide

### PR DESCRIPTION
**Pull request by memsharded**
Changelog: Feature: Allow building a compatible package still of the current profile one.
Docs: Omit

Lets keep this not documented at the moment, as there is a gap in ``conan graph build-order``

Close https://github.com/conan-io/conan/issues/16148
Close https://github.com/conan-io/conan/issues/14291

### CODE REVIEW

**Comment by memsharded** [comment]
@czoido @franramirez688 @AbrilRBS 

This PR contains both the feature + some necessary refactors to support the feature. 

I have done the refactor in its own PR to simplify the review, please check it first before reviewing this one: https://github.com/conan-io/conan/pull/16919

File: conans/client/graph/compute_pid.py [hunk 1]
```diff
@@ -79,6 +70,15 @@ def compute_package_id(node, new_config, config_version):
 
 def run_validate_package_id(conanfile):
     # IMPORTANT: This validation code must run before calling info.package_id(), to mark "invalid"
+    if hasattr(conanfile, "validate_build"):
```
Comment (memsharded) [inline conans/client/graph/compute_pid.py:73]: This is the main thing that could accidentally break something.
Previously, the ``validate_build()`` only ran for the main requested configuration. Now it is going to run for the compatible configurations, if there is some config that makes it fail other than ConanInvalidConfiguration, something might break. I think this is unlikely and we can go with this, but not impossible, but we can learn the reasons if it happens to break.

File: conans/client/graph/graph_binaries.py [hunk 2]
```diff
@@ -145,7 +145,7 @@ def _find_existing_compatible_binaries(self, node, compatibles, remotes, update)
         conanfile = node.conanfile
         original_binary = node.binary
         original_package_id = node.package_id
-
+        node.conanfile.output.info(f"Main binary package '{original_package_id}' missing")
```
Comment (AbrilRBS) [inline conans/client/graph/graph_binaries.py]:
```suggestion
        conanfile.output.info(f"Main binary package '{original_package_id}' missing")
```
Extremely minor nitpick

File: conans/client/graph/graph_binaries.py [hunk 3]
```diff
@@ -173,24 +173,41 @@ def _find_existing_compatible_binaries(self, node, compatibles, remotes, update)
...
 
+    def _find_build_compatible_binary(self, node, compatibles):
+        original_binary = node.binary
+        original_package_id = node.package_id
+        for pkg_id, compatible in compatibles.items():
+            if not compatible.cant_build:
+                node._package_id = pkg_id  # Modifying package id under the hood, FIXME
+                self._compatible_found(node.conanfile, pkg_id, compatible)
+                node.binary = BINARY_BUILD
+                return
```
Comment (AbrilRBS) [inline conans/client/graph/graph_binaries.py:188]: This means that we get to build the first one provided by the compatibility plugin, as that comes before the compatibility method? Is there a way to output some info into what changed settings are going into the compilation? Does `compatible` here not store the generated output from the `compatibility` calls somewhere?

**Review by AbrilRBS — APPROVED** [review]
This is awesome! Some comments about user feedback

File: conans/client/graph/graph_binaries.py (hunk 3 above)
Comment (memsharded) [reply conans/client/graph/graph_binaries.py:188]: Yes, we could print some message, at least the "existing binary" I think it is printing the selected compatible configuration, this should do the same, I'll add some messages.

**Review by czoido — APPROVED** [review]

**Review by memsharded — COMMENTED** [review]
What would be the behavior if some users define some ``compatibility.py`` for different compiler versions? Some might be ok and might work, like VS, but using a different ``gcc`` version for a dependency build might be challenging, Conan does not configure the environment different for different compiler versions, so it will use the ``compiler.version=<compatible-one>`` as setting, but it will still use the current environment configuration and fail?

File: test/integration/package_id/compatible_test.py [hunk 4]
```diff
@@ -448,3 +448,175 @@ def test_compatibility_msvc_and_cppstd(self):
...
+            """)
+        c.save({"conanfile.py": conanfile})
+        settings = "-s os=Windows -s compiler=gcc -s compiler.version=11 " \
+                   "-s compiler.libcxx=libstdc++11 -s compiler.cppstd=11"
+        c.run(f"create . {settings}", assert_error=True)
+        c.assert_listed_binary({"pkg/0.1": ("bb33db23c961978d08dc0cdd6bc786b45b3e5943", "Invalid")})
+        assert "pkg/0.1: Invalid: Current cppstd (11)" in c.out
+
+        c.run(f"create . {settings} --build=compatible:&")
+        print(c.out)
```
Comment (franramirez688) [inline test/integration/package_id/compatible_test.py]:
```suggestion
```

File: test/integration/package_id/compatible_test.py [hunk 5]
```diff
@@ -448,3 +448,175 @@ def test_compatibility_msvc_and_cppstd(self):
...
+            """)
+        c.save({"conanfile.py": conanfile})
+        settings = "-s os=Windows -s compiler=gcc -s compiler.version=11 " \
+                   "-s compiler.libcxx=libstdc++11  -s compiler.cppstd=11"
+        c.run(f"create . {settings}", assert_error=True)
+        c.assert_listed_binary({"pkg/0.1": ("bb33db23c961978d08dc0cdd6bc786b45b3e5943", "Invalid")})
+        assert "pkg/0.1: Invalid: Current cppstd (11)" in c.out
+
+        c.run(f"create . {settings} --build=missing", assert_error=True)
+        # the one for cppstd=17 is built!!
```
Comment (franramirez688) [inline test/integration/package_id/compatible_test.py]: Is this comment right?

**Review by franramirez688 — APPROVED** [review]
It looks great!

File: test/integration/package_id/compatible_test.py [hunk 6]
```diff
@@ -448,3 +448,174 @@ def test_compatibility_msvc_and_cppstd(self):
...
```
Comment (memsharded) [inline test/integration/package_id/compatible_test.py]:
```suggestion
```

**Pull request merged**